File: Entrada de Dados/AlfaNum/alfanum.java
import java.util.Scanner;

public class Main {
  public static boolean VALIDA_HEXA(String VALOR) {
    for (char c : VALOR.toCharArray()) {
      char CARACTERE = Character.toUpperCase(c);
      if (!(Character.isDigit(CARACTERE) || (CARACTERE >= 'A' && CARACTERE <= 'F'))) {
        return false;
      }
    }
    return true;
  }

  public static void main(String[] args) {
    Scanner entrada = new Scanner(System.in);
    System.out.println();
    String VALOR_HEX = "";
    boolean SAIR_LOOP = false;
    do {
      System.out.print("Entre valor hexadecimal de 4 caracteres: ");
      VALOR_HEX = entrada.nextLine();
      if (VALOR_HEX.length() != 4) {
        System.out.println("Valor invalido, deve ter exatamente 4 caracteres.\n");
      } else if (!VALIDA_HEXA(VALOR_HEX)) {
        System.out.println("Valor invalido, apenas caracteres [0-9] e [A-F].\n");
      } else {
        SAIR_LOOP = true;
      }
    } while (!SAIR_LOOP);
    entrada.close();
    VALOR_HEX = VALOR_HEX.toUpperCase();
    System.out.println("Valor fornecido: " + VALOR_HEX + "\n");
  }
}

File: Entrada de Dados/AlfaNum/alfanum.pas
Program LeituraHexadecimal;

Uses
  SysUtils;

Var
  VALOR_HEX: String;

Function VALIDA_HEXA(VALOR: String): Boolean;
Var
  I: integer;
  CARACTERE: char;
Begin
  For I := 1 To Length(VALOR) Do
  Begin
    CARACTERE := UpCase(VALOR[I]);
    If Not (CARACTERE In ['0'..'9', 'A'..'F']) Then
      Exit(False);
  End;
  Exit(True);
End;

Begin
  WriteLn;
  Repeat
    Write('Entre valor hexadecimal de 4 caracteres: ');
    ReadLn(VALOR_HEX);
    If (Length(VALOR_HEX) <> 4) Then
    Begin
      WriteLn('Valor invalido, deve ter exatamente 4 caracteres.');
      WriteLn;
      Continue;
    End;
    If Not (VALIDA_HEXA(VALOR_HEX)) Then
    Begin
      WriteLn('Valor invalido, apenas caracteres [0-9] e [A-F].');
      WriteLn;
      Continue;
    End;
    Break;
  Until False;
  VALOR_HEX := UpperCase(VALOR_HEX);
  WriteLn('Valor fornecido: ', VALOR_HEX);
  Writeln;
end.
